flink-runtime/src/main/java/org/apache/flink/runtime/execution/librarycache/FlinkUserCodeClassLoaders.java

```diff
@@ -82,4 +89,61 @@ public static ResolveOrder fromString(String resolveOrder) {
 			super(urls, parent);
 		}
 	}
+
+	/**
+	 * Ensures that holding a reference on the context class loader outliving the scope of user code does not prevent
+	 * the user classloader to be garbage collected (FLINK-16245).
+	 *
+	 * <p>This classloader delegates to the actual user classloader. Upon {@link #close()}, the delegate is nulled
+	 * and can be garbage collected. Additional class resolution will be resolved solely through the bootstrap
+	 * classloader and most likely result in ClassNotFound exceptions.
+	 */
+	private static class SafetyNetWrapperClassLoader extends URLClassLoader
+			implements Closeable {
+		private static final Logger LOG = LoggerFactory.getLogger(SafetyNetWrapperClassLoader.class);
+
+		private FlinkUserCodeClassLoader inner;
+
+		SafetyNetWrapperClassLoader(FlinkUserCodeClassLoader inner) {
+			super(new URL[0], null);
+			this.inner = inner;
+		}
+
+		@Override
+		public void close() {
+			if (inner != null) {
+				try {
+					inner.close();
+				} catch (IOException e) {
+					LOG.warn("Could not close user classloader", e);
+				}
+			}
+			inner = null;
+		}
+
+		@Override
+		protected Class<?> loadClass(String name, boolean resolve) throws ClassNotFoundException {
+			if (inner == null) {
+				return super.loadClass(name, resolve);
```

Comment:
Why are we not failing immediately in this case? This introduces subtle inconsistent behaviors, that to fully grasp require a good understanding of how class-loading works in the JVM which really shows in the added `testClosingOfClassloader` test. I would wager that barely anyone would be able to explain why the class cannot be loaded at the end of the test.

Reply:
I understand where you are coming from, but this will effectively make any job with leaked context classloader fail eventually, even if just basic Java stuff needs to be loaded. 
If we want to go that route, we should think twice what to throw. An `IllegalStateException` would be the correct one, but does it fit in all cases? We could also throw a `ClassNotFoundException` to not change the behavior too much and provide a good description, but that may trigger some unwanted fallback logic in user libraries.

Reply:
> any job with leaked context classloader [...] fail eventually

Given that the CL is only closed when all tasks on the TE have stopped, how could it affect a _job_? Leftover threads, sure, but what happens to those shouldn't affect job execution, and failing as early as possible might actually be a benefit.

> [throwing a CNFE] may trigger some unwanted fallback logic in user libraries

This can happen in any case; for all we now the library could be catching throwables.

I'd fail with an IllegalStateException; I think when users see a CNFE they first consider it as a class-loading issue in Flink (regardless of the error message).

Reply:
> Given that the CL is only closed when all tasks on the TE have stopped, how could it affect a job? Leftover threads, sure, but what happens to those shouldn't affect job execution, and failing as early as possible might actually be a benefit.

I actually meant it the other way around. What happens if we have job1 and job2, where the CL of job1 is leaked forward while executing job2. The TM would die taking job2 with it with a somewhat unrelated error. In particular, I'd think it would be pretty confusing/inperformant for larger batch clusters. @StephanEwen WDYT?

But since I'm also a fan of failing early, I'd go that way once Stephan agrees (I kinda had the impression that he wanted it to be graceful).

Reply:
Why would the TM die because of an exception in user-code, disconnected from any task threads no less?

Reply:
I thought that's what you meant with fail early. If you look into https://issues.apache.org/jira/browse/FLINK-11205, you can see the main motivation for this PR.

Here Stephan wrote 

> Some thoughts on how to approach this are:
> ...
> As a final safety-net, the TMs kill/restart themselves when the metaspace blows up FLINK-16225

Btw, this answers your initial point in this thread: throwing ISE it is. Now, I just need to figure out how to make sure that this exception will actually let TM fail instead of any later job.

Reply:
> I thought that's what you meant with fail early.

No, with fail early I meant that the (leaked) thread with the leaked classloader reference would crash ASAP if it tries to load a class when the CL was closed.

> Now, I just need to figure out how to make sure that this exception will actually let TM fail.

I don't see why we should do this.

FLINK-16225 aims to have the TM kill itself on a Metaspace OOM which is fine, but I find no reference anywhere that we should always fail the TM at the slightest hint of a leaked thread and/or classloader, which in any case would be out-of-scope of this PR?

Reply:
> No, with fail early I meant that the (leaked) thread with the leaked classloader reference would crash ASAP if it tries to load a class when the CL was closed.

That sounds completely fine with me, but are we sure the CL is only leaked by leaking the thread? I understood that the Apache Commons Logging framework lives outside of the thread (in app classloader) and keeps the reference to the user classloader for the whole TM lifecycle.

Reply:
I think we can be sure a class loaded from the user-jar can only leak something through a separate thread.

But we aren't sure at all what libraries in /lib are doing.
JCL for example does keep references around (and even [documents](https://cwiki.apache.org/confluence/display/COMMONS/Logging+UndeployMemoryLeak) that), but I don't see a way to handle this from our end.

My conclusions would be that users should stop throwing dependencies into /lib and create fatter user-jars instead.

Reply:
My feeling is to go with Arvid's solution. Keep the loading of classes that are not from the user code classloader intact.

Reply:
Are we hoping that these threads shutdown gracefully if we allow them to load java classes, or are we worried that crashing them this way can cause side-effects?

Reply:
My thought would be mainly understandable behavior. The change targets cases of caching beyond execution of slot related tasks, such as in logging frameworks. Users would really wonder why java core classes could not be loaded any more.

Aside from that, if the logging framwork only accesses core classes, this change would break nothing, while forbidding to load all classes might break some setups.

Related: Might make sense to wrap the `ClassNotFoundException` and enhance the error message with "Flink user code classloader is shut down".

Reply:
> Users would really wonder why java core classes could not be loaded any more.

That's a fair point.